qml/js/functions.js

```diff
@@ -406,3 +406,28 @@ function handleErrorMessage(code, message) {
         appNotification.show(message);
     }
 }
+
+function getMessagesNeededForwardPermissions(messages) {
+    var neededPermissions = ["can_send_messages"];
+
+    var mediaMessageTypes = ["messageAudio", "messageDocument", "messagePhoto", "messageVideo", "messageVideoNote", "messageVoiceNote"];
+    var otherMessageTypes = ["messageAnimation", "messageGame", "messageSticker"]
+    for(var i = 0; i < messages.length && neededPermissions.length < 3; i += 1) {
+        var type = messages[i]["content"]["@type"];
+        var permission = "";
+        if(type === "messageText") {
+            continue;
+        } else if(type === "messagePoll") {
+            permission = "can_send_polls";
+        } else if(mediaMessageTypes.indexOf(type) > -1) {
+            permission = "can_send_media_messages";
+        } else if(otherMessageTypes.indexOf(type) > -1) {
+            permission = "can_send_other_messages";
+        }
+
+        if(permission !== "" && neededPermissions.indexOf(permission) === -1) {
+            neededPermissions.push(permission);
+        }
+    }
+    return neededPermissions;
+}
```

Comment:
This function looks like a perfect candidate for being implemented in C++. Moving QVariantMaps between C++ and QML is A LOT cheaper than this sort of JavaScript code.

Reply:
Sometimes I almost forget that we're dealing with a really old and comparatively slow js engine here. ;) 
I'll try my best.

Reply:
@monich After implementing a (really simple) C++ alternative, I discovered that this computation really isn't very relevant: 

For 50 messages on my Xperia X, JS gets it done at roughly 0-1ms*/150µs**, while my admittedly pretty unoptimized c++ takes (including communication overhead) roughly 5ms* for the same messages. 

I consider forwarding > 50 messages at the same time an edge case.

Here's my ultra basic c++ for reference, even I can guess one could preallocate strings or even lists.
But with that JS performance, I don't think it's really worth the effort regardless of further c++-optimizations, at least in this case. Agreed?

```
QStringList FernschreiberUtils::getMessagesNeededForwardPermissions(const QVariantList &messages)
{
    QStringList permissions({"can_send_messages"});
    QStringList mediaMessageTypes({"messageAudio", "messageDocument", "messagePhoto", "messageVideo", "messageVideoNote", "messageVoiceNote"});
    QStringList otherMessageTypes({"messageAnimation", "messageGame", "messageSticker"});

    QString permission;

    for (QVariantList::const_iterator j = messages.begin(); j != messages.end() && permissions.count() < 3; j++)
    {
        QString messageType = j->toMap().value("content").toMap().value("@type").toString();
        if(messageType == "messageText") {
            continue;
        }
        else if(messageType == "messagePoll") {
            permission = "can_send_polls";
        }
        else if(mediaMessageTypes.contains(messageType)) {
            permission = "can_send_media_messages";
        }
        else if(otherMessageTypes.contains(messageType)) {
            permission = "can_send_other_messages";
        }
        if(!permissions.contains(permission)) {
            permissions.append(permission);
        }
    }
    return permissions;
}

```

*) Really simple Date comparison from JS, so only ms granularity.
**) QML Profiler

Reply:
Hmm... I would have expected some noticeable difference. Regarding your C++ code, commonly used `QStrings` can be statically or locally pre-allocated (or `QStringLiteral` be used) to avoid ASCII => UCS2 conversions of `char*`, but I doubt that it would be a huge saver. I guess `QVariantMap` is so inefficient that once you start using it, you're doomed in terms of performance))

If message was a pre-parsed struct/class, its type a enum and so on, I'm sure C++ would perform much better.

Oh well. I would still prefer C++ because it's easier to debug, but if you're more comfortable with JavaScript, then go ahead, of course.